### ISSUE TITLE
When called frequently, UDP parsePacket() causes access point failure

### ISSUE THREAD

**Issue by BrianG4** [arduino-libraries/WiFiNINA#117]
Setup
-
1 x MKR WiFi 1010 as WiFi access point (hub),
1 x MKR WiFi 1010 connected to ap (node).

Both Arduinos using firmware version 1.3.0 and programmed by Arduino IDE 1.8.12 on Linux.

Node sends UDP message (counter) to hub, hub receives message, updates some counters and flashes builtin LED over and over. Node sends messages as often as possible.

Timing
-
Without any restrictions, the hub's UDP server quickly stops working (< 1 minute) and causes the network to disconnect clients.

Started looking at delays because of nina-fw issue 18: https://github.com/arduino/nina-fw/issues/18 (closed in FW 1.2.4)

Adding delays to reduce frequency of checks to the UDP parsePacket() as well as adding a delay between parsePacket() and read() can improve stability. (10ms UDP check interval, 1ms delay before read() - works better, but I have not tested it exhaustively.)  I am not sure what is causing the problem and therefore not sure if this is a reasonable fix.

Sketch
-
I used the following sketch to investigate / define the problem.
The first device turned on starts the AP as hub, following device joins the network as node.

```
#include <SPI.h>
#include <WiFiNINA.h>
#include <WiFiUdp.h>

#include "arduino_secrets.h"

#define UDP_CHECK_INTERVAL 0 // check udp interval (ms)
#define UDP_DELAY_AFTER 0 // after parsePacket(), before read()

//works better with UDP_CHECK_INTERVAL = 10 && UDP_DELAY_AFTER = 1

#define UDP_SEND_PERIOD 0 // 0 = sending as fast as possible

// network
char ssid[] = SECRET_SSID;
char pass[] = SECRET_PASS;
IPAddress apIP(192, 168, 4, 1); // not explicitly set, relies on defaults
int status = WL_IDLE_STATUS;

// UDP variables (UDP:2048)
int UDP_port = 2048;
WiFiUDP Udp;
char packetBuffer[255 + 1];
char udpMsg[] = "message:";

// flasher
int led =  LED_BUILTIN;
bool led_on = false;

// timers and counters
unsigned long timer_next_UDP_msg_send;
unsigned long timer_next_UDP_chk;
unsigned int counter = 0; // message being sent from 'node'
unsigned int missed_count = 0;
unsigned int udp_msg_received_count = 0;
unsigned int udp_sending_error = 0;
unsigned long millis_started = 0; // time program started - millis() at end of setup()

enum e_wifi_mode{
    wm_hub,
    wm_node
} wifi_mode;

void setup()
{
    // leave time to clear output on Serial monitor
    delay(5000); 

    // start Serial
    Serial.begin(9600);
    while(!Serial)
    {
        ;
    }

    // Serial greeting
    Serial.println("Test UDP Bug - (q)uit to end wifi, (s)tatus");
    
    // prepare led pin as signal that UDP messages are being received
    pinMode(led, OUTPUT);
    
    // WiFi connection: if wifi exists, join existing wifi using begin(), else start wifi AP with beginAP()
    
    // check if wifi exists
    bool wifi_found = false;
    int numSSID = WiFi.scanNetworks();
    if(numSSID > 0)
    {
        for(int i = 0; i < numSSID; i++)
        {
            if(strcmp(WiFi.SSID(i), ssid) == 0)
            {
                wifi_found = true;
                Serial.println(F("existing wifi found."));
                break;
            }
        }
    }
    // start wifi
    if(wifi_found)
    {
        // connect to existing AP as 'node'
        status = WiFi.begin(ssid, pass);
        
        if(status != WL_CONNECTED)
        {
            Serial.println("Connecting to access point failed");
            while (true);// don't continue
        }
        
        Serial.println("Connected to access point");
        wifi_mode = wm_node;
    }
    else
    {
        // start AP as 'hub'
        status = WiFi.beginAP(ssid, pass);
        
        if (status != WL_AP_LISTENING)
        {
            Serial.println("Creating access point failed");
            while (true); // don't continue
        }
        
        Serial.println("Created access point");
        wifi_mode = wm_hub;
    }
    
    delay(10000);

    // start UDP Server
    Udp.begin(UDP_port);
    // initialize timers
    timer_next_UDP_msg_send = millis() + UDP_SEND_PERIOD;
    timer_next_UDP_chk = millis() + UDP_CHECK_INTERVAL;
    
    millis_started = millis(); // for print_status()
}

void loop()
{
    // check Serial for inputs ('q' or 'quit' to stop wifi; 's' or 'status' to Serial.print status report)
    check_serial();
    
    //check for wifi connections / disconnections
    check_wifi_status();
    
    // if hub - check UDP messages
    // if node - send UPD message
    if(wifi_mode == wm_hub && timer_next_UDP_chk <= millis())
    {
        //set time before running check, keeps time taken for check() out of equation
        timer_next_UDP_chk = millis() + UDP_CHECK_INTERVAL;
        // not worried about unsigned long overflow
        // (occurs after about 50 days)
        
        // check hub for incoming UDP node messages
        // and update counter
        check_udp();
    }
    else if(wifi_mode == wm_node && timer_next_UDP_msg_send <= millis())
    {
        timer_next_UDP_msg_send += UDP_SEND_PERIOD;
        
        // send udp message to ap hub
        send_udp_msg();
    }
}

void check_serial()
{
    char buffer[255];
    strcpy(buffer,"");
    int i = 0;
    while(Serial.available())
    {
        char inChar = (char)Serial.read();
        
        char newLine = '\n';
        
        if(inChar == newLine || i == 255)
        {
            buffer[i] = '\0';
            break;
        }
        else
        {
            buffer[i] = inChar;
            i++;
        }
    }
    
    /* DEBUG
    if(i > 0)
    {
        Serial.print("got: ");
        Serial.println(buffer);
    }
    //*/
    
    if(strcmp(buffer,"q") == 0 || strcmp(buffer,"quit") == 0)
    {
        Serial.println("Serial quit command received");
        quit();
    }
    else if(strcmp(buffer,"s") == 0 || strcmp(buffer,"status") == 0)
    {
        Serial.println("Serial status command received");
        print_status();
    }
}

void quit()
{
    print_status();
    WiFi.disconnect();
    WiFi.end();
    Serial.println("WiFi ended");
    while (true); // don't continue
}

void check_wifi_status()
{
    // check for change in wifi status
    // compare the previous status to the current status
    if(status != WiFi.status())
    {
        // it has changed update the variable
        status = WiFi.status();

        if(wifi_mode == wm_hub)
        {
            if(status == WL_AP_CONNECTED)
            {
                // a device has connected to the AP
                // does not say how many devices are connected
                Serial.println("a device has connected to this AP");
            }
            else if(status == WL_AP_LISTENING)
            {
                // a device has disconnected from the AP, back in listening mode
                Serial.println("all devices have disconnected from this AP");
                
                // expecting at least on device to be connected
                // if it disconnects, there has been an error - stop
                quit();
            }
            else
            {
                Serial.print("status changed to something else: ");
                Serial.println(status);
                quit();
            }
        }
        else
        {// wifi_mode == wm_node
            if(status == WL_CONNECTED)
            {
                Serial.println("connected to AP as 'node'");
            }
            else
            {
                Serial.println("disconnected from AP");
                
                // this means that the AP has failed
                // stop trying to send messages
                quit();
            }
        }
    }
}

void check_udp()
{
    
    unsigned long t1;
    unsigned long t2;

    t1 = millis();
    
    int packetSize = Udp.parsePacket();
    
    //* DEBUG
    // report time taken for parsePacket
    t2 = millis() - t1;
    if(t2 > 1)
    { // don't print everything
        Serial.print("parsePacket() ms: ");
        Serial.println(t2);
    }
    //*/
    
    if(packetSize)
    {
        
#if UDP_DELAY_AFTER > 0
        delay(UDP_DELAY_AFTER); //ADDED with reference to arduino / nina-fw issue: UDP stops working on heavy load #18 (corrected in FW 1.2.4)
#endif
        
        udp_msg_received_count++;
        
        IPAddress remoteIP = Udp.remoteIP();
        
        int len = Udp.read(packetBuffer, 255);
        
        if(len > 0)
        {
            packetBuffer[len] = 0;
        }
        
        //process message
        unsigned int old_counter = counter;
        if(len > 0 && strstr(packetBuffer, udpMsg) == packetBuffer)
        {
            char* ptr = packetBuffer + strlen(udpMsg);
            counter = atol(ptr); // update counter
            
            // some statistics
            if(counter > (old_counter + 1))
            {
                // missed something
                missed_count += (counter - old_counter) - 1;
            }
            if(counter < old_counter)
            { // message probably incomplete (truncated) - ignore
                counter = old_counter;
                
                //* DEBUG
                Serial.println("Error: message incomplete (presumed truncated)");
                //*/
            }
            
            // flash LED_BUILTIN to show it's working
            led_on = ((counter / 100) % 2 == 0);
            if(led_on)
            {
                digitalWrite(led, HIGH);
            }
            else
            {
                digitalWrite(led, LOW);
            }
        }
    }
}

void send_udp_msg()
{
    char ReplyBuffer[255];
    counter++;
    sprintf(ReplyBuffer, "%s%d", udpMsg, counter);
    Udp.beginPacket(apIP, UDP_port);
    Udp.write(ReplyBuffer);
    if(Udp.endPacket() == 0)
    {
        udp_sending_error++;
        
        /* DEBUG
        Serial.print("error sending packet: ");
        Serial.println(udp_sending_error);
        //*/
    }
}

void print_status()
{
    //time running
    int sec = (millis() - millis_started) / 1000;
    int min = sec / 60;
    int hrs = min / 60;
    int days = hrs / 24;
    Serial.println("Time Running: ");
    
    Serial.print(days);
    Serial.print(" days, ");
    Serial.print(hrs % 24);
    Serial.print(" hours, ");
    Serial.print(min % 60);
    Serial.print(" minutes, ");
    Serial.print(sec % 60);
    Serial.println(" seconds");
    
    Serial.print("UDP_CHECK_INTERVAL: ");
    Serial.print(UDP_CHECK_INTERVAL);
    Serial.print(", UDP_DELAY_AFTER: ");
    Serial.println(UDP_DELAY_AFTER);
    
    Serial.print("counter: ");
    Serial.print(counter);
    
    if(wifi_mode == wm_hub)
    {
        Serial.print(", missed_count: ");
        Serial.print(missed_count);
        Serial.print(" (");
        Serial.print(missed_count / (counter / 100));
        Serial.print("%)");
        Serial.print(", received: ");
        Serial.print(udp_msg_received_count);
    }
    else
    {
        Serial.print(", sending errors: ");
        Serial.print(udp_sending_error);
    }
    Serial.print(", WiFi.status(): ");
    Serial.println(WiFi.status());
    Serial.println();
}
```
hope this helps.

**Comment by OlliSchm** [comment]
Hey Brian, same problem here.

I have the same setup consisting of two MKR WiFi 1010. One in Ap-Mode and the other connected to it, talking over UDP Port 2390.
Both on 1.3.0. The AP also hosts a webserver.

After some time the AP crashes and has to be started again by calling `WiFi.end()` and `WiFi.beginAP()`.
With `UDP.begin(Port)` it's possible to restart the UDP-Connection. But it's not possible to restart the webServer bei calling `webServer.begin()`.

When starting another UDP-Connection or a second webServer, it takes much less time to crash.
(Btw. Arduino says the connection to the module is lost)

Excuse formal errors I am completely new here.

Best regards, Oliver

**Comment by BrianG4** [comment]
Hi Oliver,

Thanks for the message, I am new here too.

It seems there are several, possibly related issues with the wifi on the MKR1010, I have tried to isolate one example. As you pointed out, restarting the udp server does seem to extend operation, but does not overcome the issue.

The approach I am taking is the idea that there is a background task that needs an unusually long time to finish what it is doing. Therefore spacing out checks and adding delays reduces the stress on this task. I don't know if this just kicks the problem further away in time, or if it lets the task complete and achieves stability - I keep interrupting my tests.

I have encountered the problem you described with a tcp http server crashing too. I noticed that sending http messages using client.write() was more stable than using client.print() and that sending smaller strings was better (40 chars at a time). Adding some delays also seemed to help. The web pages take longer to load, but more pages can be sent.

example tcp send for char array (you could use substr if using a string)
`char string_to_send[SIZE];`
```
#if defined(ARDUINO_SAMD_MKRWIFI1010)
    int str_len = strlen(pStr);
    char* str_ptr = (char*) pStr;

    int chars_to_send = 40; // seems better limiting to 40 chars at once - still testing

    chars_to_send = (chars_to_send > str_len) ? str_len : chars_to_send;
    //only send as many as needed

    while (str_len > 0 && str_len >= chars_to_send)
    {
        client.write(str_ptr, chars_to_send);
        
        delay(1); // give time to do things...
        
        str_ptr += chars_to_send;
        str_len = strlen(str_ptr);
        chars_to_send = (chars_to_send > str_len) ? str_len : chars_to_send;
    }
#else
    //method using client.print() for other hardware
#endif
```
Best regards, Brian

**Comment by BrianG4** [comment]
Solved by [arduino/nina-fw#49](https://github.com/arduino/nina-fw/pull/49)
